### PR TITLE
[Integration][Pagerduty] - Increase PagerDuty Rate Limit Threshold to Prevent 429 Errors

### PR DESCRIPTION
**Pull request by shariff-6**
### **User description**
# Fix: Increase PagerDuty Rate Limit Threshold to Prevent 429 Errors

### What
- Increased the proactive rate limit threshold from `1` to `15` in the PagerDuty rate limiter
- Improved error handling in `send_api_request` by explicitly initializing the response variable
- Updated and added tests to verify the new threshold behavior

### Why
- The previous threshold of `1` was too aggressive, causing the integration to continue making requests until the absolute limit was reached
- With `MAX_CONCURRENT_REQUESTS = 10`, multiple concurrent requests could all proceed when `remaining > 1`, leading to rate limit violations (429 errors) before the limiter could pause
- PagerDuty enforces a strict rate limit of 960 requests per minute, and with large datasets (e.g., 10,000 incidents requiring 100+ API requests), the integration was frequently hitting this limit
- The old error handling pattern using `if "response" in locals()` was less readable and maintainable

### How
- Changed `_proactive_remaining_threshold` from `1` to `15` in `PagerDutyRateLimiter.__init__()` to provide a safety buffer for concurrent requests
- Updated `send_api_request` to explicitly initialize `response: httpx.Response | None = None` and capture `response = e.response` in the exception handler
- Replaced `if "response" in locals()` with `if response:` for cleaner, more Pythonic code
- Updated `test_enters_pause_when_remaining_le_threshold` to test the new threshold value of 15
- Added `test_does_not_pause_when_remaining_above_threshold` to verify requests proceed when remaining > 15

## Type of change

Please leave one option from the following and delete the rest:

- [x] Bug fix (non-breaking change which fixes an issue)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Completed a full resync from a freshly installed integration and it completed successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Bug fix

___

### **Description**
- Increased proactive rate limit threshold from 1 to 15 to prevent 429 errors

- Improved error handling by explicitly initializing response variable

- Updated tests to verify new threshold behavior and edge cases

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Rate Limiter Threshold"] -->|"Increased from 1 to 15"| B["Concurrent Request Buffer"]
  C["Error Handling"] -->|"Explicit initialization"| D["Response Variable"]
  E["Test Coverage"] -->|"Added new test"| F["Threshold Validation"]
  B --> G["Prevent 429 Errors"]
  D --> G
  F --> G
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>rate_limiter.py</strong><dd><code>Increase proactive rate limit threshold to 15</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/clients/rate_limiter.py

<ul><li>Changed <code>_proactive_remaining_threshold</code> from 1 to 15 to provide safety <br>buffer for concurrent requests<br> <li> Prevents rate limit violations when multiple concurrent requests <br>proceed simultaneously</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2683/files#diff-998d3e8aee5d838ba16d8344636ef7dba5006f249c3fdda4289bc4cc550e764f">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Error handling</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pagerduty.py</strong><dd><code>Improve error handling with explicit response initialization</code></dd></summary>
<hr>

integrations/pagerduty/clients/pagerduty.py

<ul><li>Explicitly initialized <code>response</code> variable to <code>None</code> at start of try block<br> <li> Captured <code>response = e.response</code> in exception handler for cleaner error <br>handling<br> <li> Replaced <code>if "response" in locals()</code> with <code>if response:</code> for more Pythonic <br>code<br> <li> Updated error logging to use captured response object</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2683/files#diff-79febc4a748111076c6ff542e5813561fd6e51a42ac805e969719e0aa9ebac24">+6/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_rate_limiter.py</strong><dd><code>Update and add rate limiter threshold tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/tests/test_rate_limiter.py

<ul><li>Renamed test from <code>test_enters_pause_when_remaining_le_1</code> to <br><code>test_enters_pause_when_remaining_le_threshold</code><br> <li> Updated test to verify pause behavior at new threshold of 15 remaining <br>requests<br> <li> Added new test <code>test_does_not_pause_when_remaining_above_threshold</code> to <br>verify requests proceed when remaining > 15</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2683/files#diff-0cda983bbb1ed376efb34e8fd4ea0169f58da830fda23d84921f09ea608ba432">+24/-3</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Add changelog entry for version 0.4.35</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/CHANGELOG.md

<ul><li>Added version 0.4.35 release notes documenting the rate limit <br>threshold increase<br> <li> Documented bug fix for preventing 429 errors with large datasets and <br>concurrency</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2683/files#diff-c45e12ac890e8ded370bbeaf51628dc3377baedbeae3a3bf083f406923b5b813">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Bump integration version to 0.4.35</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/pyproject.toml

- Bumped version from 0.4.34 to 0.4.35

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2683/files#diff-18835376b907755feaa49186a6b680ab836bd4462435f2ad4071272cf67d3b68">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: integrations/pagerduty/clients/pagerduty.py [hunk 1]
```diff
@@ -318,20 +319,22 @@ async def send_api_request(
...
 
                 logger.error(
-                    f"HTTP error for endpoint '{endpoint}': Status code {status_code}, Method: {method}, Query params: {query_params}, Response text: {e.response.text}"
+                    f"HTTP error for endpoint '{endpoint}': Status code {status_code}, Method: {method}, Query params: {query_params}, Response text: {response.text}"
                 )
                 raise
 
             finally:
-                if "response" in locals():
+                if response:
```
Comment (mk-armah) [inline integrations/pagerduty/clients/pagerduty.py]: revert

File: integrations/pagerduty/clients/rate_limiter.py [hunk 2]
```diff
@@ -47,7 +47,7 @@ def __init__(self, max_concurrent: int = 10) -> None:
         self._lock = asyncio.Lock()
 
         self.rate_limit_info: Optional[RateLimitInfo] = None
-        self._proactive_remaining_threshold = 1
+        self._proactive_remaining_threshold = 15
```
Comment (mk-armah) [inline integrations/pagerduty/clients/rate_limiter.py]: Lets do 90% of the remaining from the headers

**Review by mk-armah — CHANGES_REQUESTED** [review]

File: integrations/pagerduty/clients/rate_limiter.py
Comment (mk-armah) [inline integrations/pagerduty/clients/rate_limiter.py]:
```suggestion
        self._semaphore = asyncio.BoundedSemaphore(max_concurrent)
```

**Comment by mk-armah** [comment]
Update the date in the changelog

**Review by matan84 — APPROVED** [review]
lgtm
